File: springboot-azure/spring-webflux-cosmosdb-mongo-api-crud/src/main/java/com/knf/dev/demo/service/UserServiceImpl.java
package com.knf.dev.demo.service;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import com.knf.dev.demo.model.User;
import com.knf.dev.demo.repository.UserRepository;

import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

@Service
public class UserServiceImpl implements UserService {

    @Autowired
    private UserRepository userRepository;

    @Override
    public Mono<User> save(User user) {
        return this.userRepository.save(user);
    }

    @Override
    public Mono<User> delete(String id) {

        return this.userRepository
            .findById(id).flatMap(p ->
                this.userRepository
                    .deleteById(p.getId())
                 .thenReturn(p));

    }

    @Override
    public Mono<User> update(String id, User user) {

        return this.userRepository.findById(id)
            .flatMap(u -> {
                   u.setId(id);
                   u.setEmailId(user.getEmailId());
                   u.setName(user.getName());
                   return save(u);
                     }).switchIfEmpty(Mono.empty());
    }

    @Override
    public Flux<User> findAll() {
        return this.userRepository.findAll();
    }

    @Override
    public Mono<User> findById(String id) {
        return this.userRepository.findById(id);
    }
}

File: kotlin-spring-main/kotlin-spring-thymeleaf-login-signup/src/main/kotlin/com/knf/dev/demo/kotlinspringthymeleafloginsignup/service/UserServiceImpl.kt
package com.knf.dev.demo.kotlinspringthymeleafloginsignup.service

import com.knf.dev.demo.kotlinspringthymeleafloginsignup.dto.UserRegistrationDto
import com.knf.dev.demo.kotlinspringthymeleafloginsignup.model.Role
import com.knf.dev.demo.kotlinspringthymeleafloginsignup.model.User
import com.knf.dev.demo.kotlinspringthymeleafloginsignup.repository.UserRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.core.userdetails.UsernameNotFoundException
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import org.springframework.stereotype.Service
import java.util.*
import java.util.function.Function
import java.util.stream.Collectors


@Service
class UserServiceImpl(private val userRepository: UserRepository) : UserService {
    @Autowired
    private val passwordEncoder: BCryptPasswordEncoder? = null
    override fun save(registrationDto: UserRegistrationDto?): User? {
        val user = User(
            registrationDto!!.firstName,
            registrationDto.lastName, registrationDto.email,
            passwordEncoder!!.encode(registrationDto.password),
            Arrays.asList(Role("ROLE_USER"))
        )
        return userRepository.save(user)
    }

    @Throws(UsernameNotFoundException::class)
    override fun loadUserByUsername(username: String): UserDetails {
        val user = userRepository.findByEmail(username)
            ?: throw UsernameNotFoundException("Invalid username or password.")
        return org.springframework.security.core.userdetails.User(
            user.email, user.password,
            mapRolesToAuthorities(user.roles!!)
        )
    }

    private fun mapRolesToAuthorities(roles: Collection<Role>): Collection<GrantedAuthority?> {
        return roles.stream().map(Function<Role, SimpleGrantedAuthority?> { role: Role ->
            SimpleGrantedAuthority(
                role.name
            )
        }).collect(Collectors.toList())
    }

    override val all: List<User?>?
        get() = userRepository.findAll()

}